### ISSUE TITLE
IBSS algorithm did not converge in 100 iterations!

### ISSUE THREAD

**Issue by liqingbioinfo** [stephenslab/susieR#176]
Dear SuSiE creators: 

> I ran SuSiE on my summary statistical data and external reference LD. I got many warning messages that said, "IBSS algorithm did not converge in 100 iterations!" I first ignored such warning messages, but then found SuSiE credible sets results were somehow against intuitions (listed below). Whether the failure of SuSiE IBSS convergence result in my strange SuSiE results? If so, I really appreciate any solution to avoid the failure of IBSS convergence. 

SNP | CHR | BP | A2 | A1 | MAF | N | Z | PIP | CREDIBLE_SET
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
rs10890343 | 1 | 46157922 | C | T | 0.288325 | 85716 | -3.03604 | 1 | 3
rs6665808 | 1 | 46162283 | C | T | 0.288348 | 85716 | -3.02961 | 1 | 4
rs61784800 | 1 | 46178112 | A | G | 0.295552 | 85716 | -2.81034 | 1 | 2
rs28376548 | 1 | 46210382 | G | A | 0.286697 | 85716 | -3.04516 | 1 | 1

> rs10890343 and rs6665808 are close to each other, and they have similar Z scores as they are in high LD. Based on SuSiE's toy example, these two highly correlated SNPs should be grouped into one credible set as we have difficulties distinguishing them. However, I got two credible sets for each of them. Any elaborations are highly appreciated! 

Yours sincerely 
Leah

**Comment by gaow** [comment]
@liqingbioinfo could you try the diagnosis along the lines of [this page](https://stephenslab.github.io/susieR/articles/susierss_diagnostic.html) see if that gives you some hints about it?

**Comment by liqingbioinfo** [comment]
Hi Gao
    
> Thank you so much for your prompt reply. 


>  I read the "LD information from the reference panel" part on your page before posting my question. In your "LD information from the reference panel" part, you demonstrated that one misleading SNP had an incorrect SNP effect size sign which led to a false positive discovery. After correcting the sign of effect size of that missing leading SNP, the false positive credible set disappeared. However, in my case, I am sure that my SNPs effect sizes signs are correct, and my external reference LD information and my summary statistics are from the same ethic. Are there any other ways to avoid the failure of IBSS convergence? How about including fewer SNPs in SuSiE? 

Best regards
Leah

**Comment by gaow** [comment]
@liqingbioinfo could you post here the plot from `kriging_rss` ? Thanks!

**Comment by liqingbioinfo** [comment]
Hi Gao
    
    Thank you very much for your suggestions. I ran "kriging_rss(zscores, R_ref, n=85716)$plot" but got an error as below:

Error in `check_aesthetics()`:
! Aesthetics must be either length 1 or the same as the data (38): y

      Can I send my scores files and my LD matrices to you? So you can reproduce my results and even errors to debug? If so, what's the best email to reach you? 

Best regards
Leah

**Comment by pcarbo** [comment]
@liqingbioinfo How many SNPs are you looking at here?

**Comment by liqingbioinfo** [comment]
@pcarbo Hi Peter, I had 5,418 z scores when I got the convergence failure. I just tried to reduce it to about 1000 (split my 5k SNPs into 5 sets, each set with around 1k z scores), and the convergence failure disappeared. Does that mean I should run susie_rss with no more than 1000 z scores?

**Comment by pcarbo** [comment]
That's interesting @liqingbioinfo; it is hard to say, it is possible you got lucky and removed some SNPs causing issues.
Could you please rerun `susie_rss` with `verbose = TRUE` and share the output to the console?

**Comment by liqingbioinfo** [comment]
@pcarbo Thanks so much, Peter. I followed your advice and ran susie_rss with verbose=TRUE on my 5,418 z scores. Please find the outputs below. 

> fitted_rss <- susie_rss(zscores, R_ref, n=85716, L = 10, verbose = TRUE)
HINT: For large R or large XtX, consider installing the Rfast package for better performance.
WARNING: XtX is not symmetric; forcing XtX to be symmetric by replacing XtX with (XtX + t(XtX))/2
[1] "objective:-121564.286011201"
[1] "objective:-121420.928545254"
[1] "objective:-121256.684175818"
[1] "objective:-121088.312820567"
[1] "objective:-120918.536214462"
[1] "objective:-120748.457449151"
[1] "objective:-120578.603311599"
[1] "objective:-120409.173385859"
[1] "objective:-120240.325384438"
[1] "objective:-120072.194488068"
[1] "objective:-119904.873966087"
[1] "objective:-119738.427870526"
[1] "objective:-119572.867855749"
[1] "objective:-119407.842906286"
[1] "objective:-119241.852181439"
[1] "objective:-119074.607408203"
[1] "objective:-118907.531289341"
[1] "objective:-118742.796378205"
[1] "objective:-118578.500782208"
[1] "objective:-118414.375249951"
[1] "objective:-118250.40007027"
[1] "objective:-118087.675809122"
[1] "objective:-117925.96048741"
[1] "objective:-117765.089686481"
[1] "objective:-117604.987921771"
[1] "objective:-117445.615983111"
[1] "objective:-117286.918360643"
[1] "objective:-117128.82780346"
[1] "objective:-116971.298348918"
[1] "objective:-116814.335543441"
[1] "objective:-116657.987024549"
[1] "objective:-116502.30795693"
[1] "objective:-116347.33870226"
[1] "objective:-116193.101988764"
[1] "objective:-116039.608201594"
[1] "objective:-115886.860796443"
[1] "objective:-115734.859765478"
[1] "objective:-115583.603474728"
[1] "objective:-115433.089543085"
[1] "objective:-115283.315237552"
[1] "objective:-115134.277646092"
[1] "objective:-114985.973753259"
[1] "objective:-114838.40047636"
[1] "objective:-114691.554683881"
[1] "objective:-114545.433207362"
[1] "objective:-114400.032849364"
[1] "objective:-114255.350389157"
[1] "objective:-114111.382587815"
[1] "objective:-113968.126191819"
[1] "objective:-113825.577936552"
[1] "objective:-113683.734549082"
[1] "objective:-113542.592750684"
[1] "objective:-113402.149258683"
[1] "objective:-113262.400788854"
[1] "objective:-113123.344056457"
[1] "objective:-112984.975778081"
[1] "objective:-112847.292672682"
[1] "objective:-112710.291463202"
[1] "objective:-112573.968876956"
[1] "objective:-112438.321646848"
[1] "objective:-112303.346512506"
[1] "objective:-112169.040220658"
[1] "objective:-112035.399525871"
[1] "objective:-111902.421191271"
[1] "objective:-111770.10198895"
[1] "objective:-111638.438700569"
[1] "objective:-111507.428117674"
[1] "objective:-111377.067042135"
[1] "objective:-111247.352286537"
[1] "objective:-111118.280674407"
[1] "objective:-110989.849040374"
[1] "objective:-110862.054230276"
[1] "objective:-110734.893100701"
[1] "objective:-110608.362518161"
[1] "objective:-110482.459356388"
[1] "objective:-110357.180489992"
[1] "objective:-110232.522780571"
[1] "objective:-110108.48304331"
[1] "objective:-109985.057981263"
[1] "objective:-109862.244079211"
[1] "objective:-109740.037532675"
[1] "objective:-109618.43439218"
[1] "objective:-109497.430841572"
[1] "objective:-109377.023598933"
[1] "objective:-109257.213357407"
[1] "objective:-109138.032970216"
[1] "objective:-108856.183253484"
[1] "objective:-108072.56499724"
[1] "objective:-107779.716861665"
[1] "objective:-107611.134044919"
[1] "objective:-107442.588540823"
[1] "objective:-107274.080240477"
[1] "objective:-107105.609032486"
[1] "objective:-106937.174803695"
[1] "objective:-106768.777438831"
[1] "objective:-106600.416821207"
[1] "objective:-106432.092832491"
[1] "objective:-106263.805352197"
[1] "objective:-106095.554258855"
[1] "objective:-105927.339429416"
Warning message:
In susie_suff_stat(XtX = XtX, Xty = Xty, n = n, yty = (n - 1) *  :
  IBSS algorithm did not converge in 100 iterations!
                  Please check consistency between summary statistics and LD matrix.
                  See https://stephenslab.github.io/susieR/articles/susierss_diagnostic.html

**Comment by pcarbo** [comment]
@liqingbioinfo You can send me an email with the inputs you provided to `kriging_rss`. You can find my email address on my webpage.

**Comment by Jesson-mark** [comment]
@pcarbo Thanks for your great work to susieR. 
I'm recently using it to do fine-mapping using public GWAS summary statistics and LD from a reference panel. I just encountered the same problem as liqingbioinfo and I have performed diagnostics listed in [Diagnostic for fine-mapping with summary statistics](https://stephenslab.github.io/susieR/articles/susierss_diagnostic.html), below is lambda and output of `kriging_rss`:

```{r}
> lambda = estimate_s_rss(z_scores, ld, n=n)
> lambda
[1] 0.611342
```

![image](https://user-images.githubusercontent.com/49034603/208017222-57af665d-eb99-4929-a5f8-451743ca1872.png)

The summary statistics of GWAS have samples above 400k, and the reference panel used to calculate LD is the same ethnic population from 1000 Genome Project, which has about 600 individuals. So does this warning inform that the sample size of reference panel is not enough to calculate LD and thus cannot represent true linkage disequilibrium?

**Comment by pcarbo** [comment]
I'm sharing the kriging_rss plots from the data shared by the previous commenter @liqingbioinfo which also suggest that there are some inconsistencies between the z-scores and the LD matrix.

![screenshot](https://user-images.githubusercontent.com/1503904/208123067-b12343be-88d7-4fe4-a399-fdad817757ab.png)

**Comment by liqingbioinfo** [comment]
@pcarbo Thank you, Peter, for your great help in plotting these figures. We agreed that these figures showing expected values deviate from observed values, especially those red dots. Should I remove these red dots? Using in-sample LD is the best choice. However, I do not have individual genotype data. What can I do to improve my SuSiE results, considering I only have summary statistic data and external LD references?

**Comment by stephens999** [comment]
the bottom line is, I think, that the field does not really have a great
answer to your question.
The problem with mismatched reference is recognized, but difficult to solve
in general.
Removing the red dots might be a good start but it is hard to say it will
suffice.
The simplest fall-back is to use L=1 because that does not require LD
information (and will work well for
regions containing at most 1 causal SNP).

I will add that in the last plot none of the z scores are very big - by eye
the region looks like it could be null?

Matthew

**Comment by gaow** [comment]
There are two issues in this type of problems:

1. Simple harmonization: major/minor allele flips, strand flips
2. Mismatched summary stats and LD reference diagnosis and QC

For 1, I believe many software packages implement some features to do that eg `bigsnpr::snp_match` and also codes from LDSC related software. 

For 2, we don't know what would be a proper method to use and fix the issue (if at all). The best we can do is to identify and QC them. For example [there is a software to do that with one approach implemented](https://github.com/mkanai/slalom).

@pcarbo do you konw if there is a good streamlined workflow for these ? We have built some for in house uses since a while ago so I did not really search actively if someone else recently posted something decent to use along these lines. @liqingbioinfo @Jesson-mark your input are welcomed!

**Comment by pcarbo** [comment]
I do not know of a streamlined workflow; in the susie-rss paper in fact we made a point of mentioning that we have a diagnostic, but not an automated pipeline for "cleaning up" the summary statistics. A simple thing you can do of course is remove the SNPs that are far away from the diagonal in the kriging plot. There is also the DENTIST software, and it is possible that this software automates this more (I don't have any experience with it).

**Comment by liqingbioinfo** [comment]
@stephens999 @gaow @pcarbo Tons of thanks. I will begin by removing the red dots and ensuring my LD reference matches my GWAS summary statistics.

**Comment by PhoebeGuo97** [comment]
Hello, I'm having the same issue. I'm wondering how to remove those red dots. I don't know what SNPs corresponding to them according to condz$plot. Thanks!

**Comment by pcarbo** [comment]
@PhoebeGuo97 It isn't very well explained in the susieR package, but you would use the `conditional_dist` output from `kriging_rss`; the S1 text of the susie-rss [*PLoS Genetics* paper](https://doi.org/10.1371/journal.pgen.1010299) (see the section "Likelihood ratio for detecting allele flips") explains how to use these diagnostic statistics to identify problematic SNPs.

**Comment by PhoebeGuo97** [comment]
@pcarbo Thanks for the information! After removing red dots, I'm still getting the same warning, i.e. IBSS algorithm did not converge in 100 iterations. I'm wondering what the range of an acceptable lambda is. I have tried two LD reference panels and removed red dots, so far the smallest lambda I can get is ~0.23. Many thanks!

**Comment by gaow** [comment]
@PhoebeGuo97 i sent you an email to test out an in house pipeline we had implementing SLALMON if you could test it out then we 'll try to clean it up and share with the authors here for review. Thanks!

**Comment by PaulaLopezMarti** [comment]
Hello, I am having a similar issue as @PhoebeGuo97. I've run susie_rss() on 1399 SNPs and I got the same warning (IBSS algorithm did not converge in 100 iterations...). I built the LD matrix from the 1KG European population (same ethnicity as in the summary statistics file) using plink1.9. I've checked for allele discrepancies (with `bigsnpr::snp_match` as suggested above) and removed those conflicting SNPs, so I had a resulting set of 1399 SNPs. When running susie_rss() I get the mentioned error. Also, my lambda score is ~0.22, and when plotting the condz$plot there are no red dots (although maybe there is too much deviation between expected and observed? I don't know which would be the "accepted" limit). Thanks!
![b21bee83-f299-4cfb-976a-4b1bec5af2db](https://user-images.githubusercontent.com/57529670/218719286-346ba5bb-a1f2-4f61-90d6-3426abbbf681.png)

**Comment by pcarbo** [comment]
@PaulaLopezMarti You could try removing some more SNPs that are furthest away from the diagonal in your plot.

Also feel free to share the console output of `susie_rss` with `verbose = TRUE`.

**Comment by gaow** [comment]
The SNPs further away from the diagonal are those with small p-values. Our diagnostic model assumes small effects though. I'm not sure how well it behaves in this case when the outliers have what seems strong effect size.

**Comment by stephens999** [comment]
the diagnostic can help diagnose problems, but here it looks like the LD matrix might just not be good enough...

**Comment by PhoebeGuo97** [comment]
I just tried rerunning susie_rss using a different LD matrix calculated from 1000 Genome phase 1. This time, it converged but it seems the LD reference still doesn't match the GWAS summary statistics. I'm kind of confused. What does the warnning "IBSS algorithm did not converge in 100 iterations" mean? Thanks.
![Screenshot 2023-02-14 at 5 26 15 PM](https://user-images.githubusercontent.com/99839214/218886943-cd2a7338-c77f-469b-adfc-cb32ef894776.png)

**Comment by gaow** [comment]
> What does the warnning "IBSS algorithm did not converge in 100 iterations"

@PhoebeGuo97 it literally means that IBSS algorithm still has not found an optimal solution after 100 iterations so it stops and gives you the solution as is. There may be many reasons for convergence failure but the most likely reason is still LD and summary statistics mismatch. I was told that you are meeting with someone from my team later this week to review this. Hopefully you'll have some more insights after that. We are still testing the pipeline on our own data. So far we find that the SLALOM procedure (a simplified DENTIST method plus some ad hoc processing based on the author's own experiences) works in many cases but can still fail in a small number of GWAS regions we finemapped. It is our plan to dig a bit more and share with the SuSiE developers what we learned and see if we can come up with some recommendations & codes for the community.

**Comment by RedietNegash** [comment]
@pcarbo @liqingbioinfo,
I am encountering the same issue with the IBSS algorithm not converging in 100 iterations. I've followed the diagnostic steps and ensured that my LD matrix and GWAS summary statistics match, but I'm still getting the same error message and unexpected results.  

Below is the R script I used along with the plot I generated.

`library(susieR)
library(curl)
library(readr)


data1 <- read_csv('./scripts/filtered_chr16_pos53828066_snps_ld_test.csv')
Rin <- read_delim("data/susie/ALL_chr/ld/test_sig_locus_mt_r2.ld", delim = "\t", col_names = FALSE)


z_scores <- data1$beta / data1$se

lambda = estimate_s_rss(z_scores, Rin, n=359983)
lambda
#[1] 0.005381591

n <- 359983
condz_in=kriging_rss(z_scores, Rin, n=n)
condz_in$plot

`

![Image](https://github.com/user-attachments/assets/fb5eef91-e230-40c2-b561-95415529f977)

Has anyone found a solution or workaround for this? I would appreciate any advice.

Thanks!

**Comment by pcarbo** [comment]
@liqingbioinfo It is quite striking that you have no z-scores near zero. It appears therefore that almost all your genetic variants are (quite strongly) associated with the phenotype/trait. This is possible in principle, but quite unusual. Are you filtering the SNPs based on their association p-values before running susie?

**Comment by RedietNegash** [comment]
@pcarbo Yes, I did filter the SNPs based on their association p-values before running SuSiE. Specifically, I first filtered SNPs with a minor allele frequency  greater than 0.05 (gwas_data_df['minor_AF'] > 0.05) and then selected only those with a p-value ≤ 5e-8

**Comment by RedietNegash** [comment]
@pcarbo okay, thank you! I will try running susie as you suggested and share the results. 

Also When calculating the LD matrices using PLINK, it loaded 49,063 variants from the .bim file, but only 259 SNPs were extracted and passed QC. The LD matrices were successfully calculated. However, many SNPs from the GWAS data were still missing in the 1000 Genomes VCF data. Since SuSiE requires the LD and GWAS data to have the same shape, it won't run unless these SNPs are aligned. 

The solution I used was to exclude the missing SNPs from the GWAS dataset to match the shape of the LD matrix, but this results in most of the SNPs being filtered out. Do you have any suggestions on how to handle this or a better way to align the datasets without losing too many SNPs?

**Comment by RedietNegash** [comment]
Hi @pcarbo,  

As per your suggestion, I tested SuSiE without filtering based on p-values, but I excluded variants with smaller MAFs due to the presence of many NaNs in the LD matrix. However, I am still encountering the same issue.


here is the code I used
```
ro.r('set.seed(123)')
fit = susieR.susie_rss(
    bhat = filtered_chr16_pos53828066_snps_ld_test["beta"].values.reshape(len(R_df), 1),
    shat = filtered_chr16_pos53828066_snps_ld_test["se"].values.reshape(len(R_df), 1),
    R = R_df,
    L = 10,
    n = 359983,
    verbose=True
)
```

### **Result**  

```
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: In addition: 
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Warning message:

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: In susie_suff_stat(XtX = XtX, Xty = Xty, n = n, yty = (n - 1) *  :
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 
 
WARNING:rpy2.rinterface_lib.callbacks:R[write to console]:  IBSS algorithm did not converge in 100 iterations!
                  Please check consistency between summary statistics and LD matrix.
                  See https://stephenslab.github.io/susieR/articles/susierss_diagnostic.html

WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: HINT: For large R or large XtX, consider installing the Rfast package for better performance.

[1] "objective:-495073.585457801"
[1] "objective:-480114.67284869"
[1] "objective:-464231.060730647"
[1] "objective:-448147.592071923"
[1] "objective:-432193.225643745"
[1] "objective:-416261.469843408"
[1] "objective:-400405.60189626"
[1] "objective:-384619.978268979"
[1] "objective:-368897.746336303"
[1] "objective:-353234.33410298"
[1] "objective:-337626.79206323"
[1] "objective:-322073.125712916"
[1] "objective:-306571.91830695"
[1] "objective:-291122.111659176"
[1] "objective:-275722.873996838"
[1] "objective:-260373.517146163"
[1] "objective:-245073.442697967"
[1] "objective:-229822.105993343"
[1] "objective:-214618.991625427"
[1] "objective:-199463.596757252"
[1] "objective:-184355.41996406"
[1] "objective:-169293.95408599"
[1] "objective:-154278.682045091"
[1] "objective:-139309.07484847"
[1] "objective:-124384.591183074"
[1] "objective:-109504.678150126"
[1] "objective:-94668.7727706759"
[1] "objective:-79876.3039708601"
[1] "objective:-65126.6948495397"
[1] "objective:-50419.3650429163"
[1] "objective:-35753.7330665223"
[1] "objective:-21129.2185516843"
[1] "objective:-6545.24431470106"
[1] "objective:7998.76178137888"
[1] "objective:22503.3651899658"
[1] "objective:36969.1232748678"
[1] "objective:51396.5841278817"
[1] "objective:65786.28560892"
[1] "objective:80138.754589026"
[1] "objective:94454.506376431"
[1] "objective:108734.044305138"
[1] "objective:122977.859479106"
[1] "objective:137186.430630516"
[1] "objective:151360.224102596"
[1] "objective:165499.693912926"
[1] "objective:179605.281897814"
[1] "objective:193677.417918769"
[1] "objective:207716.520119746"
[1] "objective:221722.995220953"
[1] "objective:235697.238846013"
[1] "objective:249639.635866988"
[1] "objective:263550.560764594"
[1] "objective:277430.377998261"
[1] "objective:291279.442375767"
[1] "objective:305098.099427681"
[1] "objective:318886.685773515"
[1] "objective:332645.529482891"
[1] "objective:346374.950426892"
[1] "objective:360075.260626015"
[1] "objective:373746.764574705"
[1] "objective:387389.759562729"
[1] "objective:401004.53598249"
[1] "objective:414591.377623228"
[1] "objective:428150.561950182"
[1] "objective:441682.360373274"
[1] "objective:455187.038502712"
[1] "objective:468664.856390706"
[1] "objective:482116.068762997"
[1] "objective:495540.925238061"
[1] "objective:508939.670533324"
[1] "objective:522312.544664323"
[1] "objective:535659.783131953"
[1] "objective:548981.617095471"
[1] "objective:562278.273545258"
[1] "objective:575549.975459429"
[1] "objective:588796.941954259"
[1] "objective:602019.388425983"
[1] "objective:615217.526689218"
[1] "objective:628391.565102402"
[1] "objective:641541.708690597"
[1] "objective:654668.159261799"
[1] "objective:667771.115514829"
[1] "objective:680850.773142026"
[1] "objective:693907.324932694"
[1] "objective:706940.960863138"
[1] "objective:719951.868185655"
[1] "objective:732940.231516093"
[1] "objective:745906.232912286"
[1] "objective:758850.051952341"
[1] "objective:771771.865807131"
[1] "objective:784671.849310999"
[1] "objective:797550.175027448"
[1] "objective:810407.013314313"
[1] "objective:823242.53238261"
[1] "objective:836056.898355874"
[1] "objective:848850.275325761"
[1] "objective:861622.825403815"
[1] "objective:874374.708774041"
[1] "objective:887106.083741761"
[1] "objective:899817.106775067"

```

![Image](https://github.com/user-attachments/assets/ee7eddbd-eeff-4898-abe1-42888526fa06)

Despite including all available variants, increasing the number of iterations, reducing L and removing missing SNPs to match the LD matrix, I am obtaining the same results.  



**Data Sources**

**GWAS Summary Statistics**:  The GWAS summary statistics are from the UK Biobank [google sheet](https://docs.google.com/spreadsheets/d/1kvPoupSzsSFBNSztMzl04xMoSC3Kcx3CrjVf4yBmESU/edit?gid=227859291#gid=227859291) with phenotype code 21001_raw for both sexes.

**Reference VCF** : [1000 Genomes Project](https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/) ``` EUR``` population

Would appreciate any insights on what might be causing this!  

Thanks,

**Comment by RedietNegash** [comment]
@pcarbo Thank you for your helpful suggestions. In my case The issue turned out to be due to the LD matrix not being positive semi-definite (PSD) which caused problems with the SuSiE model fitting and led to the IBSS algorithm not converging. 
To fix this I computed the eigenvalues of the LD matrix and found that the minimum eigenvalue was negative. To make the matrix PSD I added a small constant (epsilon) to the diagonal elements of the matrix . After this adjustments SuSiE was able to run without any convergence issues.
